subprojects/core/src/test/groovy/org/gradle/internal/vfs/HierarchicalFileWatcherUpdaterTest.groovy

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.vfs
+
+import net.rubygrapefruit.platform.file.FileWatcher
+import org.gradle.internal.vfs.watch.FileWatcherUpdater
+
+class HierarchicalFileWatcherUpdaterTest extends AbstractFileWatcherUpdaterTest {
+
+    @Override
+    FileWatcherUpdater createUpdater(FileWatcher watcher) {
+        new HierarchicalFileWatcherUpdater(watcher)
+    }
+
+    def "only adds watches for the roots of must watch directories"() {
+        def mustWatchDirectoryRoots = ["first", "second"].collect { file(it).createDir() }
+        def mustWatchDirectories = mustWatchDirectoryRoots + file("first/within").createDir()
+
+        when:
+        updater.updateMustWatchDirectories(mustWatchDirectories)
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, mustWatchDirectoryRoots) })
+        0 * _
+    }
```

Comment:
Do we need a test for "upgrading" a watch, i.e. when `first/within` is watched first, and then we need to start `first`, we'd need to unregister `first/within`, right?

Reply:
I added a test for the use-case you are describing.

Reply:
I think this is fine, though it would be more readable for me if these new cases were added as separate tests instead.

Reply:
I did split the test now.